### PR TITLE
feat(core): generate new workspaces without tasks runner options

### PR DESCRIPTION
**Pull request by AgentEnder**
## Current Behavior
New workspaces are created with a default `tasksRunnerOptions` object

## Expected Behavior
New workspaces use the new config options for configuring the default tasks runner.

## Related Issue(s)

Fixes #

### CODE REVIEW

File: packages/workspace/src/generators/ci-workflow/ci-workflow.ts [hunk 1]
```diff
@@ -22,9 +22,11 @@ export async function ciWorkflowGenerator(host: Tree, schema: Schema) {
   const options = normalizeOptions(schema);
 
   const nxJson: NxJsonConfiguration = readJson(host, 'nx.json');
-  const nxCloudUsed = Object.values(nxJson.tasksRunnerOptions).find(
-    (r) => r.runner == '@nrwl/nx-cloud' || r.runner == 'nx-cloud'
-  );
+  const nxCloudUsed =
```
Comment (FrozenPandaz) [inline packages/workspace/src/generators/ci-workflow/ci-workflow.ts:25]: Please move to a separate PR

File: packages/workspace/src/generators/ci-workflow/ci-workflow.spec.ts [hunk 2]
```diff
@@ -111,11 +111,7 @@ function setNxCloud(tree: Tree) {
   updateJson<NxJsonConfiguration>(tree, 'nx.json', (json) => {
     return {
       ...json,
-      tasksRunnerOptions: {
-        default: {
-          runner: 'nx-cloud',
-        },
-      },
+      nxCloudAccessToken: 'xxxx-xxx-xxxx',
```
Comment (FrozenPandaz) [inline packages/workspace/src/generators/ci-workflow/ci-workflow.spec.ts:114]: Please move to a separate PR

File: packages/nx/src/tasks-runner/run-command.ts [hunk 3]
```diff
@@ -474,6 +474,10 @@ function getRunnerOptions(
     result.url ??= nxJson.nxCloudUrl;
   }
 
+  if (nxJson.nxCloudEncryptionKey && isCloudDefault) {
```
Comment (FrozenPandaz) [inline packages/nx/src/tasks-runner/run-command.ts:477]: Move to a separate PR

File: packages/nx/src/tasks-runner/run-command.ts [hunk 4]
```diff
@@ -483,7 +487,10 @@ function getRunnerOptions(
   }
 
   if (defaultCacheableOperations.length) {
-    result.cacheableOperations ??= defaultCacheableOperations;
+    result.cacheableOperations ??= [];
+    result.cacheableOperations = result.cacheableOperations.concat(
```
Comment (FrozenPandaz) [inline packages/nx/src/tasks-runner/run-command.ts:491]: Move to a separate PR

File: packages/nx/src/generators/utils/deprecated.ts [hunk 5]
```diff
@@ -25,49 +25,51 @@ export function updateWorkspaceConfiguration(
...
     extends: ext,
+    generators,
+    implicitDependencies,
     installation,
-    release,
+    namedInputs,
+    npmScope,
     nxCloudAccessToken,
-    cacheDirectory,
+    nxCloudEncryptionKey,
```
Comment (FrozenPandaz) [inline packages/nx/src/generators/utils/deprecated.ts:63]: Move to a separate PR

File: packages/nx/src/daemon/client/client.ts [hunk 6]
```diff
@@ -365,7 +365,7 @@ export class DaemonClient {
     this._out = await open(DAEMON_OUTPUT_LOG_FILE, 'a');
     this._err = await open(DAEMON_OUTPUT_LOG_FILE, 'a');
 
-    if (this.nxJson.tasksRunnerOptions.default?.options?.useParcelWatcher) {
+    if (this.nxJson.tasksRunnerOptions?.default?.options?.useParcelWatcher) {
```
Comment (FrozenPandaz) [inline packages/nx/src/daemon/client/client.ts:368]: Move to a separate PR

File: packages/nx/src/config/nx-json.ts [hunk 7]
```diff
@@ -213,6 +213,11 @@ export interface NxJsonConfiguration<T = '*' | string[]> {
    */
   nxCloudUrl?: string;
 
+  /**
+   * Specifies the encryption key used to encrypt artifacts data before sending it to nx cloud.
+   */
+  nxCloudEncryptionKey?: string;
```
Comment (FrozenPandaz) [inline packages/nx/src/config/nx-json.ts:219]: Move to a separate PR

File: packages/nx/src/command-line/connect/connect-to-nx-cloud.ts [hunk 8]
```diff
@@ -14,9 +14,9 @@ export async function connectToNxCloudIfExplicitlyAsked(opts: {
 }): Promise<void> {
   if (opts['cloud'] === true) {
     const nxJson = readNxJson();
-    const runners = Object.values(nxJson.tasksRunnerOptions);
+    const runners = Object.values(nxJson.tasksRunnerOptions ?? {});
     const onlyDefaultRunnerIsUsed =
-      runners.length === 1 && runners[0].runner === 'nx/tasks-runners/default';
+      runners.length <= 1 && runners[0].runner === 'nx/tasks-runners/default';
```
Comment (FrozenPandaz) [inline packages/nx/src/command-line/connect/connect-to-nx-cloud.ts:19]: Move this to a separate PR

File: packages/nx/src/adapter/compat.ts [hunk 9]
```diff
@@ -69,6 +69,7 @@ export const allowedWorkspaceExtensions = [
   'release',
   'nxCloudAccessToken',
   'nxCloudUrl',
+  'nxCloudEncryptionKey',
```
Comment (FrozenPandaz) [inline packages/nx/src/adapter/compat.ts:72]: Move this to a separate PR

File: packages/nx/schemas/nx-schema.json [hunk 10]
```diff
@@ -90,6 +90,10 @@
       "type": "string",
       "description": "Specifies the url pointing to an instance of nx cloud. Used for remote caching and displaying run links."
     },
+    "nxCloudEncryptionKey": {
```
Comment (FrozenPandaz) [inline packages/nx/schemas/nx-schema.json:93]: Move this to a separate PR

File: packages/nx/schemas/nx-schema.json [hunk 11]
```diff
@@ -371,6 +375,10 @@
               }
             ]
           }
+        },
+        "cache": {
```
Comment (FrozenPandaz) [inline packages/nx/schemas/nx-schema.json:379]: Move this to a separate PR

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

**Review by mandarini — APPROVED** [review]
I'll make the Storybook fixes after this PR is merged.

File: packages/angular/src/generators/ng-add/migrators/migrator.ts [hunk 12]
```diff
@@ -97,8 +97,7 @@ export abstract class Migrator {
 
     const nxJson = readNxJson(this.tree);
 
-    Object.keys(nxJson.tasksRunnerOptions ?? {}).forEach((taskRunnerName) => {
-      const taskRunner = nxJson.tasksRunnerOptions[taskRunnerName];
+    Object.values(nxJson.tasksRunnerOptions ?? {}).forEach((taskRunner) => {
```
Comment (leosvelperez) [inline packages/angular/src/generators/ng-add/migrators/migrator.ts:100]: If I'm understanding the changes correctly, this shouldn't be needed anymore. This migrates an Angular CLI workspace to an Nx workspace. So, we shouldn't have `tasksRunnerOptions`.

File: packages/angular/src/generators/setup-ssr/lib/update-project-config.ts [hunk 13]
```diff
@@ -87,6 +87,11 @@ export function updateProjectConfig(tree: Tree, schema: Schema) {
     ];
     updateNxJson(tree, nxJson);
   }
+  if (nxJson.targetDefaults) {
+    nxJson.targetDefaults.server ??= {};
+    nxJson.targetDefaults.server.cache ??= true;
+    updateNxJson(tree, nxJson);
+  }
```
Comment (leosvelperez) [inline packages/angular/src/generators/setup-ssr/lib/update-project-config.ts:94]: If I look at the changes here (and in other files of this PR), if we have tasks runner options, we are going to add the target to its cacheable operations, but we are also going to have the target defaults with the `cache` option set.

Is it intentional to duplicate the configuration? Shouldn't we only have one or the other while we transition? I'd expect we:

- Set the `cache` option in target defaults for new workspaces
- Continue using `cacheableOperations` in old workspaces until we migrate them off of it

File: packages/nx/src/command-line/init/implementation/angular/standalone-workspace.ts [hunk 14]
```diff
@@ -100,6 +100,7 @@ function createNxJson(
     nxJson.targetDefaults.build = {
       dependsOn: ['^build'],
       inputs: ['production', '^production'],
+      cache: true,
```
Comment (leosvelperez) [inline packages/nx/src/command-line/init/implementation/angular/standalone-workspace.ts:103]: This shouldn't overwrite what the `createNxJsonFile` helper (updated in this PR) already sets based on the cacheable operations.

File: packages/nx/src/command-line/init/implementation/angular/standalone-workspace.ts [hunk 15]
```diff
@@ -110,7 +111,7 @@ function createNxJson(
     if (fileExists(join(repoRoot, 'karma.conf.js'))) {
       inputs.push('{workspaceRoot}/karma.conf.js');
     }
-    nxJson.targetDefaults.test = { inputs };
+    nxJson.targetDefaults.test = { inputs, cache: true };
```
Comment (leosvelperez) [inline packages/nx/src/command-line/init/implementation/angular/standalone-workspace.ts:114]: This shouldn't overwrite what the `createNxJsonFile` helper (updated in this PR) already sets based on the cacheable operations.

File: packages/nx/src/command-line/init/implementation/angular/standalone-workspace.ts [hunk 16]
```diff
@@ -120,10 +121,13 @@ function createNxJson(
     if (fileExists(join(repoRoot, 'eslint.config.js'))) {
       inputs.push('{workspaceRoot}/eslint.config.js');
     }
-    nxJson.targetDefaults.lint = { inputs };
+    nxJson.targetDefaults.lint = { inputs, cache: true };
```
Comment (leosvelperez) [inline packages/nx/src/command-line/init/implementation/angular/standalone-workspace.ts:124]: This shouldn't overwrite what the `createNxJsonFile` helper (updated in this PR) already sets based on the cacheable operations.

File: packages/nx/src/command-line/init/implementation/angular/standalone-workspace.ts [hunk 17]
```diff
@@ -120,10 +121,13 @@ function createNxJson(
...
   }
   if (workspaceTargets.includes('e2e')) {
-    nxJson.targetDefaults.e2e = { inputs: ['default', '^production'] };
+    nxJson.targetDefaults.e2e = {
+      inputs: ['default', '^production'],
+      cache: true,
```
Comment (leosvelperez) [inline packages/nx/src/command-line/init/implementation/angular/standalone-workspace.ts:129]: This shouldn't overwrite what the `createNxJsonFile` helper (updated in this PR) already sets based on the cacheable operations.

**Review by leosvelperez — CHANGES_REQUESTED** [review]

File: packages/storybook/src/migrations/update-14-1-8/change-storybook-targets-generator.spec.ts [hunk 18]
```diff
@@ -8,7 +8,9 @@ import * as extraOptions from './test-configs/extra-options-for-storybook.json';
 import * as noStorybookBuildTarget from './test-configs/no-build-storybook-target.json';
 import * as noStorybook from './test-configs/no-storybook-targets.json';
 
-describe('Change the Storybook targets for Angular projects to use native Storybooke executor', () => {
+// @TODO (@mandarini) Fixup and re-enable these tests. The snapshots are all the same currently,
+// and likely need to check **project** config, not nx.json config.
+xdescribe('Change the Storybook targets for Angular projects to use native Storybooke executor', () => {
```
Comment (Coly010) [inline packages/storybook/src/migrations/update-14-1-8/change-storybook-targets-generator.spec.ts:13]: An alternative to the TODO here would just be to delete the migrations for 14.1.8.

They're less likely to be needed in Nx 17.

If a user is on a version < 14.1.8, they should be migrating to Nx 15 first anyway, which will contain the migrations.

**Review by Coly010 — APPROVED** [review]

File: packages/angular/src/generators/setup-ssr/lib/update-project-config.ts (hunk 13 above)
Comment (FrozenPandaz) [inline packages/angular/src/generators/setup-ssr/lib/update-project-config.ts:94]:
```suggestion
    nxJson.targetDefaults ??= {};
    nxJson.targetDefaults.server ??= {};
    nxJson.targetDefaults.server.cache ??= true;
    updateNxJson(tree, nxJson);
```

File: packages/next/src/generators/custom-server/custom-server.ts [hunk 19]
```diff
@@ -102,14 +102,16 @@ export async function customServerGenerator(
...
-      json.tasksRunnerOptions.default ??= { options: {} };
-      json.tasksRunnerOptions.default.options.cacheableOperations = [
-        ...json.tasksRunnerOptions.default.options.cacheableOperations,
-        'build-custom-server',
-      ];
+      json.tasksRunnerOptions.default.options.cacheableOperations.push(
+        'build-custom-server'
+      );
+    }
+    if (json.targetDefaults) {
```
Comment (FrozenPandaz) [inline packages/next/src/generators/custom-server/custom-server.ts:112]: Always set `targetDefaults`

File: packages/nx/src/command-line/connect/connect-to-nx-cloud.ts [hunk 20]
```diff
@@ -14,9 +14,9 @@ export async function connectToNxCloudIfExplicitlyAsked(opts: {
...
     if (!onlyDefaultRunnerIsUsed) return;
```
Comment (FrozenPandaz) [inline packages/nx/src/command-line/connect/connect-to-nx-cloud.ts:20]: This logic is wrong when `taskRunnerOptions` is not defined.

The `length` of the values will be 0.

File: packages/nx/src/command-line/migrate/migrate.ts [hunk 21]
```diff
@@ -1336,7 +1336,7 @@ function showConnectToCloudMessage() {
     const nxJson = readJsonFile<NxJsonConfiguration>('nx.json');
     const defaultRunnerIsUsed =
       !nxJson.tasksRunnerOptions ||
```
Comment (FrozenPandaz) [inline packages/nx/src/command-line/migrate/migrate.ts:1338]: We also need to check they are not already using cloud

File: packages/nx/src/utils/nx-cloud-utils.ts [hunk 22]
```diff
@@ -3,27 +3,28 @@ import { NxJsonConfiguration, readNxJson } from '../config/nx-json';
...
-  const taskRunner = isNxCloudUsed(readNxJson());
-  if (!taskRunner) throw new Error('nx-cloud runner not find in nx.json');
-  return (
-    (typeof taskRunner === 'object' ? taskRunner.options.url : null) ??
-    'https://nx.app'
+export function getNxCloudUrl(nxJson: NxJsonConfiguration): string {
+  const cloudRunner = Object.values(nxJson.tasksRunnerOptions ?? {}).find(
+    (r) => r.runner == '@nrwl/nx-cloud' || r.runner == 'nx-cloud'
   );
+  if (!cloudRunner && !nxJson.nxCloudUrl)
```
Comment (FrozenPandaz) [inline packages/nx/src/utils/nx-cloud-utils.ts:16]: Check if they have the access token set

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

File: packages/angular/src/generators/setup-ssr/lib/update-project-config.ts (hunk 13 above)
Comment (AgentEnder) [reply packages/angular/src/generators/setup-ssr/lib/update-project-config.ts:94]: Jason and I talked about this, there are a few reasons that it makes sense to keep cacheableOperations up to date, and we have a migration for existing repos as is so most folks shouldn't see the duplicate implementation.

File: packages/storybook/src/migrations/update-14-1-8/change-storybook-targets-generator.spec.ts (hunk 18 above)
Comment (AgentEnder) [reply packages/storybook/src/migrations/update-14-1-8/change-storybook-targets-generator.spec.ts:13]: Yeah, that should probably happen but will be outside this PR 😄

File: packages/nx/src/command-line/connect/connect-to-nx-cloud.ts [hunk 23]
```diff
@@ -8,16 +8,29 @@ import {
   isNxCloudUsed,
 } from '../../utils/nx-cloud-utils';
 import { runNxSync } from '../../utils/child-process';
+import { NxJsonConfiguration } from '../../config/nx-json';
+
+export function onlyDefaultRunnerIsUsed(nxJson: NxJsonConfiguration) {
+  if (!nxJson.tasksRunnerOptions) {
+    if (!nxJson.nxCloudAccessToken) return false;
+    return true;
+  }
```
Comment (FrozenPandaz) [inline packages/nx/src/command-line/connect/connect-to-nx-cloud.ts:17]: This seems backwards..

They are using the default task runner if they don't have `taskRunnerOptions` AND they don't have the `nxCloudAccessToken`.

File: packages/nx/src/command-line/connect/connect-to-nx-cloud.ts [hunk 24]
```diff
@@ -8,16 +8,29 @@ import {
...
+  const defaultRunner = nxJson.tasksRunnerOptions?.default;
+  if (
+    !defaultRunner.runner ||
+    defaultRunner.runner === 'nx/tasks-runners/default'
+  ) {
+    return true;
+  }
+  return false;
```
Comment (FrozenPandaz) [inline packages/nx/src/command-line/connect/connect-to-nx-cloud.ts:25]: This is missing logic to check if they ONLY have the default configured.

EVERY task runner options value should be the default task runner.

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

**Review by FrozenPandaz — APPROVED** [review]

**Review by leosvelperez — APPROVED** [review]

File: packages/storybook/src/migrations/update-14-1-8/change-storybook-targets-generator.spec.ts (hunk 18 above)
Comment (mandarini) [reply packages/storybook/src/migrations/update-14-1-8/change-storybook-targets-generator.spec.ts:13]: Yep, I will remove that migration!

Comment (mandarini) [reply packages/storybook/src/migrations/update-14-1-8/change-storybook-targets-generator.spec.ts:13]: Oh! https://github.com/nrwl/nx/pull/19552 Jason already did!

File: packages/nx/src/command-line/connect/connect-to-nx-cloud.ts (hunk 24 above)
Comment (AgentEnder) [reply packages/nx/src/command-line/connect/connect-to-nx-cloud.ts:28]: Jason and I talked about this one offline, cloud only cares about the default runner anyways.

**Comment by mandarini** [comment]
[Storybook fixes](https://github.com/nrwl/nx/pull/19483#discussion_r1352874854) not needed any more, since old migrations were removed in PR https://github.com/nrwl/nx/pull/19552
